src/data_traits.rs

```diff
@@ -766,3 +766,16 @@ impl<'a, A: 'a, B: 'a> RawDataSubst<B> for CowRepr<'a, A> {
         }
     }
 }
+
+/// Plug the data element type of one owned array representation into another.
+///
+/// For example, `<OwnedRepr<f64> as Plug<f32>>::Type` has type `OwnedRepr<f32>`.
+pub trait Plug<A> {
```

Comment:
For the way this is being used, I think it would have to be an unsafe trait because implementing it for representation with incompatible memory layouts, the transmutes based on this would unsound?

Reply:
(Personally, since this is not a general exposition, would also give the trait a named geared towards its usage here, something like `SameRepr<A>`.)

Reply:
This trait looks similar to the existing RawDataSubst.

Reply:
@adamreichold, could you expound on this a little more?  I think you are talking about marking the `Plug` trait as unsafe to implement because someone could inadvertently use the wrong outer type in the implementation.  For example:

```rust
impl <A, B> Plug<A> for OwnedRepr<B> {
    type Type = OwnedArcRepr<A>; // Wrong! Should be OwnedRepr
}
```

Depending on the mistake, the conversion on [line 2627](https://github.com/rust-ndarray/ndarray/commit/2ac3c9f0600e10d8b6113a4270027ad816fb3849#diff-ad89e759e86f5d18eb412398d9e76b1b32b25139284f34e5d873e341ba485a96R2627) might still compile and the transmutation on [line 2632](https://github.com/rust-ndarray/ndarray/commit/2ac3c9f0600e10d8b6113a4270027ad816fb3849#diff-ad89e759e86f5d18eb412398d9e76b1b32b25139284f34e5d873e341ba485a96R2632) would be unsound.

Maybe instead of making `Plug` unsafe I can add a runtime assertion that the plugged type is equal to `T`.  This assertion would fail if `Plug` is implemented incorrectly as above.

```rust
assert!(core::any::TypeId::of::<<T as Plug<A>>::Type>() == core::any::TypeId::of::<T>());
```

Reply:
> I think you are talking about marking the Plug trait as unsafe to implement because someone could inadvertently use the wrong outer type in the implementation. 

Exactly.

> Maybe instead of making Plug unsafe I can add a runtime assertion that the plugged type is equal to T. This assertion would fail if Plug is implemented incorrectly as above.

I think you could actually use `RawDataSubst` instead of adding a new trait? At least it does seem to provide the relevant implementations for `OwnedRepr` and `OwnedArcRepr`.

Reply:
OK, I think everything has been addressed in [the latest push](https://github.com/rust-ndarray/ndarray/commit/3373ea94e54ff7ea8ddd8e987db3aad384b84627).

Reply:
The comment on encyclopedic knowledge I'm still wondering about what it means. I'm the original author of ndarray and have written almost every trait, including RawDataSubst :slightly_smiling_face: (Would the comment have been made with that knowledge? don't know.) Knowing every part of the crate would be my default position, but in recent time I've fallen off that track. The intention is to have multiple maintainers and drivers of the project.